### PR TITLE
feat: Make `hive_partitioning` parameter default to `None`, which is automatically enabled for single directory inputs, and disabled otherwise

### PR DESCRIPTION
**Pull request by nameexhaustion**

### CODE REVIEW

File: crates/polars-io/src/options.rs [hunk 1]
```diff
@@ -16,15 +16,18 @@ pub struct RowIndex {
 #[derive(Clone, Debug, Eq, PartialEq, Hash)]
 #[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
 pub struct HiveOptions {
-    pub enabled: bool,
+    /// This can be `None` to automatically enable for single directory scans
+    /// and disable otherwise. However it should be initialized if it is inside
+    /// a DSL / IR plan.
+    pub enabled: Option<bool>,
```
Comment (stinodego) [inline crates/polars-io/src/options.rs:22]: Maybe this should become an Enum like we did for `JoinCoalesce`? Not sure about best practices here.

Comment (ritchie46) [reply crates/polars-io/src/options.rs:22]: Could be both. Not super important to me (this saves somes serde bloat), the other is more readable in matching.

In any case. I am going to merge this so we can work towards the RC. :)

**Review by ritchie46 — APPROVED** [review]

**Comment by ritchie46** [comment]
Hmm... There seems to be a failing test @nameexhaustion. Not entirely sure it is related to my rebase.

**Comment by nameexhaustion** [comment]
was a bug when the directory path contained 1 file 😅

**Comment by stinodego** [comment]
@nameexhaustion FYI this is not breaking, because the default behavior for file inputs (the only kind of input that was supported in the previous release) remains unchanged.

**Comment by nameexhaustion** [comment]
But for glob patterns, in the current releases hive parsing is enabled by default, but with the next release they need to add explicitly `hive_partitioning=True` to retain the old behavior - wouldn't it be breaking there?

**Comment by stinodego** [comment]
> But in the current releases hive parsing is enabled by default, but with the next release they need to add explicitly `hive_partitioning=True` to retain the old behavior - wouldn't it be breaking there?

Ahh, my bad. It changed from True to None. I thought it defaulted to False before. Nevermind!

**Pull request merged**